### PR TITLE
Fix a couple of apartment state issues

### PR DESCRIPTION
**Pull request by kouvel**
Fix for https://github.com/dotnet/coreclr/issues/17822
- The apartment state now defaults to MTA for the main thread along with a CoInitialize
- Calling `Thread.SetApartmentState` with STA now fails as expected (different behavior from previous netcore, same behavior as netfx)

Fix for https://github.com/dotnet/coreclr/issues/17787
- `WaitHandle.WaitAll` for multiple handles is not supported on an STA thread due to issues described in https://github.com/dotnet/coreclr/issues/17787#issuecomment-385117537
- It now throws `NotSupportedException` as expected (different behavior from previous netcore, same behavior as netfx)

Fix for https://github.com/dotnet/coreclr/issues/19225

### CODE REVIEW

File: tests/src/baseservices/threading/ApartmentState/SetApartmentStateDefaultMta.cs [hunk 1]
```diff
@@ -0,0 +1,50 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Threading;
+
+public static class SetApartmentStateDefaultMtaTests
+{
+    public static int Main()
```
Comment (jkotas) [inline tests/src/baseservices/threading/ApartmentState/SetApartmentStateDefaultMta.cs]: We have tests for the STAThread/MTAThread attributes in CoreFX. They will need to be updated. I do not think we need to duplicate them here.

**Comment by jkotas** [comment]
cc @AaronRobinsonMSFT

File: tests/src/baseservices/threading/ApartmentState/SetApartmentStateDefaultMta.cs (hunk 1 above)
Comment (kouvel) [reply tests/src/baseservices/threading/ApartmentState/SetApartmentStateDefaultMta.cs]: I see, I have updated the CoreFX tests in https://github.com/dotnet/corefx/pull/31682

File: src/vm/assembly.cpp [hunk 2]
```diff
@@ -1798,11 +1798,7 @@ INT32 Assembly::ExecuteMainMethod(PTRARRAYREF *stringArgs, BOOL waitForOtherThre
 
                 Thread::ApartmentState state = Thread::AS_Unknown;
                 state = SystemDomain::GetEntryPointThreadAptState(pMeth->GetMDImport(), pMeth->GetMemberDef());
```
[the remark this answers is not in the document]
Comment (kouvel) [reply src/vm/assembly.cpp]: That function checks for the `STAThread` and `MTAThread` attributes and if one is not specified, returns `AS_Unknown`. That's expected and just means that a specific apartment state hasn't been requested. In that case, we default to MTA as though the `MTAThread` attribute was specified.

File: src/vm/comwaithandle.cpp [hunk 3]
```diff
@@ -227,12 +227,14 @@ FCIMPL4(INT32, WaitHandleNative::CorWaitMultipleNative, Object* waitObjectsUNSAF
...
-    // Perhaps in a future release we can fix this, if we aren't quite so concerned about
-    // compatibility....
+#ifdef FEATURE_COMINTEROP_APARTMENT_SUPPORT
+    // There are some issues with wait-all from an STA thread
+    // - https://github.com/dotnet/coreclr/issues/17787#issuecomment-385117537
+    if (waitForAll && numWaiters > 1 && pThread->GetApartment() == Thread::AS_InSTA)
+    {
+        COMPlusThrow(kNotSupportedException, W("NotSupported_WaitAllSTAThread"));
+    }
+#endif // FEATURE_COMINTEROP && !FEATURE_CORECLR
```
[the remark this answers is not in the document]
Comment (kouvel) [reply src/vm/comwaithandle.cpp]: Oops fixed

File: src/vm/appdomain.cpp [hunk 4]
```diff
@@ -3074,8 +3074,9 @@ void SystemDomain::SetThreadAptState (Thread::ApartmentState state)
         Thread::ApartmentState pState = pThread->SetApartment(Thread::AS_InSTA, TRUE);
         _ASSERTE(pState == Thread::AS_InSTA);
     }
-    else if (state == Thread::AS_InMTA)
+    else
     {
+        // If an apartment state was not explicitly requested, default to MTA
```
[the remark this answers is not in the document]
Comment (kouvel) [reply src/vm/appdomain.cpp]: The state here may be `AS_InMTA` or `AS_Unknown` (the only other legal values for the enum) where unknown indicates that neither `STAThread` nor `MTAThread` attributes were specified. Unknown is a valid state and we default to MTA in that case. The thread's state flags would be updated shortly after this to indicate MTA.

**Comment by kouvel** [comment]
Windows CoreFX jobs: ApartmentState_NoAttributePresent_STA_Windows_Core is an invalid test and is removed in https://github.com/dotnet/corefx/pull/31682

**Comment by kouvel** [comment]
@dotnet-bot test OSX10.12 x64 Debug CoreFX Tests

**Comment by jkotas** [comment]
>Windows CoreFX jobs: ApartmentState_NoAttributePresent_STA_Windows_Core is an invalid tes

Add it to https://github.com/dotnet/coreclr/blob/master/tests/CoreFX/CoreFX.issues.json to make the CI green

**Pull request merged**